### PR TITLE
SAK-47677 - Support endDateTime in lineitems

### PR DESCRIPTION
**Pull request by csev**

### CODE REVIEW

File: basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java [hunk 1]
```diff
@@ -1248,6 +1248,78 @@ public static String getISO8601(Date date) {
...
+	// https://www.imsglobal.org/spec/lti-ags/v2p0/#startdatetime
+	// https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html
+	// https://stackoverflow.com/questions/4024544/how-to-parse-dates-in-multiple-formats-using-simpledateformat
+	public static Date parseIMS8601(String timestamp) {
+
+		if ( timestamp == null ) return null;
+
+		// Make sure that ISO8601 Z format dates are *perfect*
+		SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
+		df.setTimeZone(TimeZone.getTimeZone("GMT"));
```
Comment (ottenhoff) [inline basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java:1274]: this seems wrong? isn't the zone coming over on the line above?

File: basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java [hunk 2]
```diff
@@ -1248,6 +1248,78 @@ public static String getISO8601(Date date) {
...
+                    "yyyy-MM-dd'T'HH:mm:ss.SSSzzzz",
+                    "yyyy-MM-dd'T'HH:mm:sszzzz",
+                    "yyyy-MM-dd'T'HH:mm:ss z",
+                    "yyyy-MM-dd'T'HH:mm:ssz",
+                    "yyyy-MM-dd'T'HH:mm:ss",
+                    "yyyy-MM-dd'T'HHmmss.SSSz",
+                    "yyyy-MM-dd",
+                    "yyyyMMdd",
+                    "dd/MM/yy",
+                    "dd/MM/yyyy"
```
Comment (ottenhoff) [inline basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java:1308]: might as well add Euro style!

MM/dd/yyyy

File: basiclti/tsugi-util/src/test/org/tsugi/basiclti/BasicLTIUtilTest.java [hunk 3]
```diff
@@ -97,12 +97,41 @@ public void testOAuthGET()
 	public void iso8601Text() throws Exception {
 		String x = BasicLTIUtil.getISO8601();
 		assertTrue(x.contains("Z"));
+		assertTrue(x.contains("T"));
 		String target = "2017-08-20:10:00:00";
+		String iso8601 = "2017-01-20T10:00:00Z";
```
Comment (ottenhoff) [inline basiclti/tsugi-util/src/test/org/tsugi/basiclti/BasicLTIUtilTest.java:102]: should try out the standard form:

2022-08-13T11:01:41+00:00

2022-08-13T11:01:41-04:00

Comment (csev) [reply basiclti/tsugi-util/src/test/org/tsugi/basiclti/BasicLTIUtilTest.java:102]: I do test these forms in the loop variation of the test.

            "2017-01-20T11:00:00+01:00",
            "2017-01-20T09:00:00-01:00",

File: basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java (hunk 2 above)
Comment (csev) [reply basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java:1308]: In my testing, the order of these strings matter :)  If something is vague like

01/01/2010

The first one that succeeds in the list wins.   If on the other hand it is clear like

12-25-2010

Then the dd/MM/yyyy will fail to convert and the MM/dd/yyy will work. :)  No way to win all the time :)

**Pull request merged**